smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php

```diff
@@ -0,0 +1,277 @@
+<?php
+
+namespace Drupal\smaily_for_drupal\Form;
+
+use Drupal\Core\Form\ConfigFormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\HtmlCommand;
+use Drupal\Core\Ajax\ReplaceCommand;
+
+/**
+ * Smaily module admin configuraiton form.
+ */
+class SmailyForDrupalAdminForm extends ConfigFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEditableConfigNames() {
+    return [
+      'smaily_for_drupal.adminsettings',
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'smaily_for_drupal_admin_form';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('smaily_for_drupal.adminsettings');
+
+    // Wrapper for editing and returning form data with ajax.
+    $form['smaily_container'] = [
+      '#type' => 'container',
+      '#attributes' => [
+        'id' => 'smaily-wrapper',
+      ],
+    ];
+
+    $form['smaily_container']['credentials'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('API credentials'),
+    ];
+
+    $form['smaily_container']['credentials']['message'] = [
+      '#type' => 'markup',
+      '#markup' => '<div class="smaily_message"></div>',
+    ];
+
+    $form['smaily_container']['credentials']['smaily_domain'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Domain'),
+      '#default_value' => $config->get('smaily_domain', ''),
+      '#attributes' => [
+        'placeholder' => $this->t('For example "demo" from https://demo.sendsmaily.net/'),
+      ],
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['credentials']['smaily_username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Username'),
+      '#default_value' => $config->get('smaily_username', ''),
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['credentials']['smaily_password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $config->get('smaily_password', ''),
+      '#required' => TRUE,
+
+    ];
+
+    $form['smaily_container']['credentials']['validate_credentials'] = [
+      '#type' => 'button',
+      '#value' => $this->t('Validate credentials'),
+      '#ajax' => [
+        'event' => 'click',
+        'callback' => '::validateCredentials',
+        'wrapper' => 'smaily-wrapper',
+      ],
+      // Limit errors here, validateCredentials will highlight errors in autoresponder/title.
+      '#limit_validation_errors' => [],
+    ];
+
+    $username = $config->get('smaily_username');
+    $password = $config->get('smaily_password');
+    $domain = $config->get('smaily_domain');
+
+    // Fetch and update autoresponders every time form is built.
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
```

Comment:
What happens if `smaily_domain` does not have a value?

Reply:
cURL will say: Could not resolve host: .sendsmaily.net

Reply:
It would be better to not make any request if any of the API credentials are missing. It is a great way of increasing efficiency.

Reply:
The validation would stop any half-filled api credentials from being saved, even with Ajax.
Is it a just-in-case scenario? I could only see this happening when someone manually edits the mySQL table

Reply:
OK. Maybe I don't understand how and when `buildForm` is called. Doesn't it get called when the form is rendered in the admin panel? If it does, wouldn't it still call `fetchAutoresponders` even if no settings have been saved before?

Reply:
I added an !empty($domain) check (now for all credentials).
It would still call `fetchAutoresponders` but the `if` check wouldn't start a cURL session.